Fix xnnpack quantization discrepancy for non-fp32

jackzhxng opened the pull request:
## Summary
Perform quantization on the weights expressed in their original dtype (from the checkpoint) by performing source transformations before dtype cast. Previously the model was being converted to the `dtype_override` arg's dtype and then quantized. This eliminates supposedly eliminates quantization noise.

Note - no need to worry about https://github.com/pytorch/ao/blob/main/torchao/quantization/GPTQ.py#L1168, precision is passed in with the checkpoint dtype

### Comparison of arbitrary q_proj tensor from sample Llama checkpoint:
Before:
```
Mismatched elements: 3260378 / 4194304 (77.7%)
Greatest absolute difference: 0.08802086114883423 at index (1129, 604) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 1350) (up to 1.3e-06 allowed)
Signal-to-noise: 32.8974 dB
```

After: no difference

## Test plan

### Manual testing
```
python -m examples.models.llama.export_llama \
-v -c xl_consolidated/consolidated_renamed.pth \
-p xl_consolidated/et_params.json -kv -d fp32 \
-qmode 8da4w --group_size 32 -X \
--use_sdpa_with_kv_cache \
--output_name quantized_baseline.pte \
--max_context_length 4096 -E 4,32
```

With the following inserted after the quantization:

```
edge_manager.model(
    torch.tensor([[2, 3, 4]], dtype=torch.long),
    {"input_pos": torch.tensor([0], dtype=torch.long)},
)
```

And the following modifications to GPTQ.py in torchao: https://github.com/pytorch/ao/pull/1756 for testing.

jackzhxng commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -561,43 +563,73 @@ def _prepare_for_llama_export(args) -> LLMEdgeManager:
+    edge_manager = edge_manager.set_output_dir(output_dir_path).source_transform(
+        _get_source_transforms(
+            args.model, DType.from_torch_dtype(checkpoint_dtype), args
         )
```
Added
```
edge_manager.model(
    torch.tensor([[2, 3, 4]], dtype=torch.long),
    {"input_pos": torch.tensor([0], dtype=torch.long)},
)
```

Here to test

jackzhxng commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -561,43 +563,73 @@ def _prepare_for_llama_export(args) -> LLMEdgeManager:
+    # We want to do compute the actual ops in the precision of the dtype_override,
+    # since the precision of the quantized linear will initially be the dtype of the
+    # checkpoint, not the dtype_override.
+    def _set_precision_to_fp32(module):
```
cc @kimishpatel for the issue we were discussing

kimishpatel commented on examples/models/llama/source_transformation/quantize.py:
```diff
@@ -827,4 +828,28 @@ def _load_torchao_aten_lib(libname):
+        if isinstance(child, Int8DynActInt4WeightLinear):
+            # Change the precision attribute to 'fp32'
+            child.precision = dtype
+            print(f"Changed precision of {name} to {dtype}")
```
use logging

kimishpatel commented on examples/models/llama/model.py:
```diff
@@ -277,10 +287,12 @@ def __init__(self, **kwargs):
 
     def get_eager_model(self) -> torch.nn.Module:
-        if self.dtype:
+        return self.model_
```
is this a typo?

jackzhxng replied:
Ah I was going to remove everything below and just have `return self.model_`

kimishpatel commented on examples/models/llama/model.py:
```diff
@@ -171,7 +172,16 @@ def __init__(self, **kwargs):
         # Within the device="meta" context, tensors that are created do not carry data.
         # They possess all other metadata a tensor carries such as size, stride, requires_grad.
         with torch.device("meta"):
+            # Model itself is loaded in default dtype, fp32.
```
why is this the case?

kimishpatel commented on examples/models/llama/model.py:
```diff
@@ -171,7 +172,16 @@ def __init__(self, **kwargs):
+            if self.dtype:
+                self.model_.to(dtype=self.dtype)
+
+            # Convert the model's weights only to the checkpoint's dtype, so that
+            # the checkpoint can be loaded into the model's state dict in its
+            # own dtype w/o potential precision loss.
+            for param in self.model_.parameters():
+                param.data = param.data.to(dtype=self.checkpoint_dtype)
```
We shouldnt have to do this if the checkpoint is directly loaded, no? Not sure whats happening with `self.model_.to(....`

kimishpatel commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -563,43 +564,55 @@ def _prepare_for_llama_export(args) -> LLMEdgeManager:
+    )
+
+    # Assumes the checkpoint has uniform dtype.
+    checkpoint_dtype = next(edge_manager.model.parameters()).dtype
```
I thought you already stored `checkpoint_dtype` in the model

kimishpatel commented on examples/models/llama/export_llama_lib.py:
```diff
@@ -563,43 +564,55 @@ def _prepare_for_llama_export(args) -> LLMEdgeManager:
-        .source_transform(_get_source_transforms(args.model, dtype_override, args))
     )
 
+    _set_quantized_computation_dtype(
```
besides the changes you are doing in this function, dont you also need to do `edge_manager.model.to(self.dtype)`?

jackzhxng committed:
Fix xnnpack quantization discrepancy for non-fp32 (#8488)

Summary:
Perform quantization on the weights expressed in their original dtype (from the checkpoint) by passing in the checkpoint dtype to the quantization source transformation and modifying the computation dtype (the result dtype of the dequant, the dtype that the ops are actually computed in) to the dtype override. We must do it this way since the checkpoint and computation dtype are coupled into a single `precision` parameter in the torchao api, and that is something that we cannot change.

Note - no need to worry about https://github.com/pytorch/ao/blob/main/torchao/quantization/GPTQ.py#L1168, precision is passed in with the checkpoint dtype

### Comparison of arbitrary q_proj tensor from sample Llama checkpoint:
Before:
```
Mismatched elements: 3260378 / 4194304 (77.7%)
Greatest absolute difference: 0.08802086114883423 at index (1129, 604) (up to 1e-05 allowed)
Greatest relative difference: 1.0 at index (0, 1350) (up to 1.3e-06 allowed)
Signal-to-noise: 32.8974 dB
```

After: no difference


Test Plan:
### Manual testing
```
python -m examples.models.llama.export_llama \
-v -c xl_consolidated/consolidated_renamed.pth \
-p xl_consolidated/et_params.json -kv -d fp32 \
-qmode 8da4w --group_size 32 -X \
--use_sdpa_with_kv_cache \
--output_name quantized_baseline.pte \
--max_context_length 4096 -E 4,32
```

With the following inserted after the quantization:

```
edge_manager.model(
    torch.tensor([[2, 3, 4]], dtype=torch.long),
    {"input_pos": torch.tensor([0], dtype=torch.long)},
)
```

And the following modifications to GPTQ.py in torchao: https://github.com/pytorch/ao/pull/1756 for testing.

### Automated testing
+ existing CI tests

### Regression testing
TBD

Reviewed By: kimishpatel

Differential Revision: D70184325

Pulled By: jackzhxng
